module/zfs/dmu_object.c

```diff
@@ -58,83 +70,114 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 		ASSERT3S(dn_slots, <=, DNODE_MAX_SLOTS);
 	}
 
-	mutex_enter(&os->os_obj_lock);
+	/*
+	 * The "chunk" of dnodes that is assigned to a CPU-specific
+	 * allocator needs to be at least one block's worth, to avoid
+	 * lock contention on the dbuf.  It can be at most one L1 block's
+	 * worth, so that the "rescan after polishing off a L1's worth"
+	 * logic below will be sure to kick in.
+	 */
+	if (dnodes_per_chunk < DNODES_PER_BLOCK)
+		dnodes_per_chunk = DNODES_PER_BLOCK;
+	if (dnodes_per_chunk > L1_dnode_count)
+		dnodes_per_chunk = L1_dnode_count;
+
+	object = atomic_inc_64_nv(cpuobj);
```

Comment:
I thought about that. That's the way it previously worked with `os->os_object_next`. My concern was that if we don't atomically update `*cpuobj` here then it becomes possible for concurrent threads to enter the for loop having the same value in `object`. But I guess we are still protected in that case by taking `dn_struct_rwlock` below.